doc/barclamp.md

```diff
@@ -0,0 +1,1177 @@
+verbosely or tersely as you like (please put at least one complete sentence in
+there).
+
+#### Recipes
+
+Location: `chef/cookbooks/mybarclamp/recipes/*.rb`
+
+[Recipes](https://docs.chef.io/recipes.html) make up the bulk of your chef
+cookbook's payload. These recipes are written in Ruby (hence they must have a
+`.rb` extension) and are the smallest organization unit of a chef cookbook.
+They consist of [resource](https://docs.chef.io/resource.html) definitions that
+define various aspects of system state, such as a configuration file's content,
+packages to install, or a service to enable and start. In this step the recipes
+do not need to contain any code, yet. You primarily need to know which recipes
+to define and what each recipe's purpose is at this stage.
+
+You are free to organize your recipes as you see fit. There are no
+hard-and-fast rules for this since they mostly depend on the application your
+barclamp deploys:
+
+For instance, if your application (like most OpenStack components) consists of
+an API service and a RabbitMQ driven backend service, you might want to split
+these into a `api.rb` and `backend.rb` service. Likewise, if you need an
+agent to run on compute nodes, this might warrant a dedicated `agent.rb`
+recipe. If you want to use Crowbar's high availability features, it might also
+make sense to have a dedicated `ha.rb` recipe for users deploying it in a
+highly available manner.
+
+On a more general note, feel free to take a look at the other cookbooks'
+recipes in `chef/cookbooks/` for inspiration. We have barclamps for a range of
+OpenStack services already and one of these might just match your own
+application's architecture perfectly.
+
+#### Roles
+
+Locations:
+
+* Role recipes: `chef/cookbooks/mybarclamp/recipes`
+* Role definitions: `chef/roles/role_mybarclamp_*.rb`
+
+Once you've got a set of chef recipes, you'll need to organize them into roles.
+A role aggregates one or more chef recipes and deploys all of its component
+chef recipes. Crowbar can assign roles to one or more nodes (subject to
+constraints defined in the Crowbar application). In the HA case it can assign
+roles to one or more clusters of roles. It pays to spend some time on getting
+the roles you need right from the start because you will need to modify a lot
+of code to add new roles or remove existing ones later. Roles vary in scope
+from deploying just a single service (such as a logging agent) to deploying
+multiple related services (such as all of an OpenStack service's controller
+side daemons). All roles' unifying feature is that they are a container for
+something that gets deployed to a certain type of machine which may among other
+things be:
+
+* "Any machine orchestrated by Crowbar"
+
+* "Any OpenStack controller node"
+
+* "Any OpenStack compute node"
+
+* "Any Node in the Ceph Cluster"
+
+* "Any Ceph OSD node"
+
+We have a limited range of node types in Crowbar to define types of roles.
+Beyond that
```

Comment:
Ah, thanks. I left that for "later" because I first needed to look up a bunch of URLs. Guess I forgot about it later...fixed.